iceoryx2-services/discovery/src/service_discovery/service.rs

```diff
@@ -0,0 +1,417 @@
+
+    /// The maximum number of samples subscribers are permitted to hold loans for.
+    pub max_borrrowed_samples: usize,
+
+    /// The number of older samples the subscriber can request from the service when starting.
+    pub history_size: usize,
+
+    /// Whether to send notifications on changes.
+    pub send_notifications: bool,
+
+    /// The maximum number of listeners to the service permitted.
+    pub max_listeners: usize,
+}
+
+impl Default for Config {
+    fn default() -> Self {
+        Self {
+            sync_on_initialization: true,
+            include_internal: true,
+            publish_events: true,
+            max_subscribers: 10,
+            max_buffer_size: 10,
+            history_size: 10,
+            max_borrrowed_samples: 10,
+            send_notifications: true,
+            max_listeners: 10,
+        }
+    }
+}
+
+/// The service discovery service.
+///
+/// This service is responsible for tracking and publishing information about
+/// services in the system. It can detect when services are added or removed,
+/// and notify interested parties about these changes.
+///
+/// # Type Parameters
+///
+/// * `S` - The service type that this discovery service operates on.
+pub struct Service<S: ServiceType> {
+    discovery_config: Config,
+    iceoryx_config: IceoryxConfig,
+    _node: Node<S>,
+    publisher: Option<Publisher<S, [u8], ()>>,
+    notifier: Option<Notifier<S>>,
+    tracker: Tracker<S>,
+}
+
+impl<S: ServiceType> Service<S> {
+    /// Creates the service discovery service.
+    ///
+    /// This function initializes a new service discovery service with the provided
+    /// configuration. The service will track services of type `S` in the system.
+    ///
+    /// # Parameters
+    ///
+    /// * `discovery_config` - Configuration for the discovery service.
+    /// * `iceoryx_config` - Configuration for the underlying iceoryx system.
+    ///
+    /// # Returns
+    ///
+    /// A result containing either the created service or an error if creation failed.
+    pub fn create(
+        discovery_config: &Config,
+        iceoryx_config: &IceoryxConfig,
+    ) -> Result<Self, CreationError> {
+        let node = NodeBuilder::new()
+            .config(iceoryx_config)
+            .create::<S>()
+            .map_err(|_| CreationError::NodeCreationFailure)?;
+
+        let mut publisher = None;
+        if discovery_config.publish_events {
+            let publish_subscribe = node
+                .service_builder(service_name())
+                .publish_subscribe::<Payload>()
+                .subscriber_max_buffer_size(discovery_config.max_buffer_size)
+                .subscriber_max_borrowed_samples(discovery_config.max_borrrowed_samples)
+                .history_size(discovery_config.history_size)
+                .max_subscribers(discovery_config.max_subscribers)
+                .max_publishers(1)
+                .open_or_create()
+                .map_err(CreationError::from)?;
```

Comment:
This can be simplified

```suggestion
                .open_or_create()?;
```

Same further below

Reply:
Squashed into: https://github.com/eclipse-iceoryx/iceoryx2/pull/703/commits/5b00835df9c6492c24b690d1f8c888074a43bab2